azure/services/privatelinks/spec.go

```diff
@@ -0,0 +1,239 @@
+			}
+		}
+		ipConfiguration := network.PrivateLinkServiceIPConfiguration{
+			Name: pointer.String(fmt.Sprintf("%s-natipconfig-%d", natIPConfiguration.Subnet, i+1)),
+			PrivateLinkServiceIPConfigurationProperties: &network.PrivateLinkServiceIPConfigurationProperties{
+				Subnet: &network.Subnet{
+					ID: pointer.String(
+						fmt.Sprintf(
+							"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s",
+							s.SubscriptionID,
+							s.ResourceGroup,
+							s.VNet,
+							natIPConfiguration.Subnet)),
+				},
+				PrivateIPAllocationMethod: ipAllocationMethod,
+				PrivateIPAddress:          privateIpAddress,
+			},
+		}
+		ipConfigurations = append(ipConfigurations, ipConfiguration)
+		ipConfigurations[0].Primary = pointer.Bool(true)
+	}
+
+	// Load balancer front-end IP configurations
+	var frontendIPConfigurations []network.FrontendIPConfiguration
+	for _, frontendIPConfigName := range s.LBFrontendIPConfigNames {
+		frontendIPConfig := network.FrontendIPConfiguration{
+			ID: pointer.String(
+				fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/loadBalancers/%s/frontendIPConfigurations/%s",
+					s.SubscriptionID,
+					s.ResourceGroupName(),
+					s.LoadBalancerName,
+					frontendIPConfigName)),
+		}
+		frontendIPConfigurations = append(frontendIPConfigurations, frontendIPConfig)
+	}
+
+	privateLinkToCreate := network.PrivateLinkService{
+		Name:     pointer.String(s.Name),
+		Location: pointer.String(s.Location),
+		PrivateLinkServiceProperties: &network.PrivateLinkServiceProperties{
+			IPConfigurations:                     &ipConfigurations,
+			LoadBalancerFrontendIPConfigurations: &frontendIPConfigurations,
+			EnableProxyProtocol:                  s.EnableProxyProtocol,
+		},
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: s.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Name:        pointer.String(s.Name),
+			Additional:  s.AdditionalTags,
+		})),
+	}
+
+	if len(s.AllowedSubscriptions) > 0 {
+		privateLinkToCreate.Visibility = &network.PrivateLinkServicePropertiesVisibility{
+			Subscriptions: &s.AllowedSubscriptions,
+		}
+	}
+	if len(s.AutoApprovedSubscriptions) > 0 {
+		privateLinkToCreate.AutoApproval = &network.PrivateLinkServicePropertiesAutoApproval{
+			Subscriptions: &s.AutoApprovedSubscriptions,
+		}
+	}
+
+	if existing != nil {
+		existingPrivateLink, ok := existing.(network.PrivateLinkService)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.PrivateLinkService", existing)
+		}
+
+		if isExistingUpToDate(existingPrivateLink, privateLinkToCreate) {
+			return nil, nil
+		}
+	}
+
+	return privateLinkToCreate, nil
+}
+
+func isExistingUpToDate(existing network.PrivateLinkService, wanted network.PrivateLinkService) bool {
+	// NAT IP configuration is not checked as it cannot be changed.
+
+	// Check load balancer configurations
+	var wantedFrontendIDs []string
```

Comment:
FYI, I believe the PR is blocked on the code coverage tests passing

Reply:
Code coverage tests should be fine now https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3747#issuecomment-1697387584.